libtenzir/builtins/formats/feather2.cpp

```diff
@@ -0,0 +1,203 @@
+    auto decode_result
+      = stream_decoder.Consume(as_arrow_buffer(std::move(payload)));
+    if (!decode_result.ok()) {
+      diagnostic::error("failed to decode the byte stream into a record batch")
+        .note("{}", decode_result.ToString())
+        .emit(ctrl.diagnostics());
+      co_return;
+    }
+
+    while (!listener->record_batch_buffer.empty()) {
+      auto batch = listener->record_batch_buffer.front();
+      listener->record_batch_buffer.pop();
+      auto validate_status = batch->Validate();
+      TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+      co_yield table_slice(batch);
+    }
+
+    if (done) {
+      TENZIR_INFO("returning");
+      co_return;
+    }
+  }
+}
+
+auto print_feather(
+  table_slice input, operator_control_plane& ctrl,
+  const std::shared_ptr<arrow::ipc::RecordBatchWriter>& stream_writer,
+  const std::shared_ptr<arrow::io::BufferOutputStream>& sink)
+  -> generator<chunk_ptr> {
+  auto batch = to_record_batch(input);
+  auto validate_status = batch->Validate();
+  TENZIR_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+
+  auto stream_writer_status = stream_writer->WriteRecordBatch(*batch);
+  if (!stream_writer_status.ok()) {
+    diagnostic::error("failed to write a record batch to the stream")
+      .note("{}", stream_writer_status.ToString())
+      .emit(ctrl.diagnostics());
+  }
+  auto finished_buffer_result = sink->Finish();
+  if (!finished_buffer_result.ok()) {
+    diagnostic::error("failed to close the stream and return the buffer")
+      .note("{}", finished_buffer_result.status().ToString())
+      .emit(ctrl.diagnostics());
+    co_return;
+  }
+  auto finished_buffer = finished_buffer_result.ValueOrDie();
+  ;
+  auto chunk = chunk::make(finished_buffer);
+  co_yield chunk;
+  auto reset_buffer_result
+    = sink->Reset(); // the buffer is reinit with newly allocated memory because
+                     // the API does not offer a Reset that just clears the
+                     // original data
+  if (!reset_buffer_result.ok()) {
+    diagnostic::error("failed to reset buffer")
+      .note("{}", reset_buffer_result.ToString())
+      .emit(ctrl.diagnostics());
+  }
+}
+
+// questions: should be include logic to ensure entire record batch recorded ()
+
+class feather2_parser final : public plugin_parser {
+public:
+  feather2_parser() = default;
+  auto name() const -> std::string override {
+    return "feather2";
+  }
+
+  auto
+  instantiate(generator<chunk_ptr> input, operator_control_plane& ctrl) const
+    -> std::optional<generator<table_slice>> override {
+    return parse_feather(std::move(input), ctrl);
+  }
+
+  friend auto inspect(auto& f, feather2_parser& x) -> bool {
+    return f.object(x).fields();
+  }
+};
+
+class feather2_printer final : public plugin_printer {
+public:
+  feather2_printer() = default;
+
+  auto name() const -> std::string override {
+    // FIXME: Rename this and the file to just feather.
```

Comment:
Please do this to all filenames and code after rebasing onto main after #4072 is merged.

Reply:
Given that #4072 turned out to be a lot more complicated than we had originally anticipated I ended up pushing the remaining changes to this PR and closed the other one.